pkg/hub/addon/discovery_controller.go

```diff
@@ -110,48 +112,68 @@ func (c *addOnFeatureDiscoveryController) sync(ctx context.Context, syncCtx fact
 func (c *addOnFeatureDiscoveryController) syncAddOn(ctx context.Context, clusterName, addOnName string) error {
 	klog.V(4).Infof("Reconciling addOn %q", addOnName)
 
-	labels := map[string]string{}
+	cluster, err := c.clusterLister.Get(clusterName)
+	if errors.IsNotFound(err) {
+		// no cluster, it could be deleted
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("unable to find cluster with name %q: %w", clusterName, err)
+	}
+	// no work if cluster is deleting
+	if !cluster.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	labels := cluster.Labels
+	if len(labels) == 0 {
+		labels = map[string]string{}
+	}
+
+	// make a copy to check if the cluster labels are changed
+	labelsCopy := map[string]string{}
+	for key, value := range labels {
+		labelsCopy[key] = value
+	}
+
 	addOn, err := c.addOnLister.ManagedClusterAddOns(clusterName).Get(addOnName)
 	switch {
 	case errors.IsNotFound(err):
 		// addon is deleted
-		key := fmt.Sprintf("%s%s-", addOnFeaturePrefix, addOnName)
-		labels[key] = ""
+		key := fmt.Sprintf("%s%s", addOnFeaturePrefix, addOnName)
+		delete(labels, key)
```

Comment:
should this be labels[key] = nil?

Reply:
looks like it will not work since labels are known type(`map[string]string`), set value to `nil`, it will complain;
```bash
pkg/hub/addon/discovery_controller.go:144:17: cannot use nil as string value in assignment
pkg/hub/addon/discovery_controller.go:148:17: cannot use nil as string value in assignment
pkg/hub/addon/discovery_controller.go:219:23: cannot use nil as string value in assignment
pkg/hub/addon/discovery_controller.go:234:23: cannot use nil as string value in assignment
```